classes/citation/Repository.php

```diff
@@ -0,0 +1,288 @@
+        return $id;
+    }
+
+    /** @copydoc DAO::update() */
+    public function edit(Citation $citation, array $params): void
+    {
+        $newCitation = clone $citation;
+        $newCitation->setAllData(array_merge($newCitation->_data, $params));
+        Hook::call('Citation::edit', [$newCitation, $citation, $params]);
+        $this->dao->update($newCitation);
+    }
+
+    /** @copydoc DAO::delete() */
+    public function delete(Citation $citation): void
+    {
+        Hook::call('Citation::delete::before', [$citation]);
+        $this->dao->delete($citation);
+        Hook::call('Citation::delete', [$citation]);
+    }
+
+    /**
+     * Delete a collection of citations
+     */
+    public function deleteMany(Collector $collector): void
+    {
+        foreach ($collector->getMany() as $citation) {
+            $this->delete($citation);
+        }
+    }
+
+    /**
+     * Insert on duplicate update.
+     */
+    public function updateOrInsert(Citation $citation): int
+    {
+        return $this->dao->updateOrInsert($citation);
+    }
+
+    /**
+     * Get all citations for a given publication.
+     *
+     * @return array<Citation>
+     */
+    public function getByPublicationId(int $publicationId): array
+    {
+        return $this->getCollector()
+            ->filterByPublicationId($publicationId)
+            ->orderBy(Collector::ORDERBY_SEQUENCE)
+            ->getMany()
+            ->all();
+    }
+
+    /**
+     * Get all raw citations for a given publication.
+     */
+    public function getRawCitationsByPublicationId(int $publicationId): Collection
+    {
+        return $this->dao->getRawCitationsByPublicationId($publicationId);
+    }
+
+    /**
+     * Delete a publication's citations.
+     */
+    public function deleteByPublicationId(int $publicationId): void
+    {
+        $this->dao->deleteByPublicationId($publicationId);
+    }
+
+    /**
+     * Import and replace citations from a raw citation list of the particular publication.
+     *
+     * @hook Citation::importCitations::before [[$publicationId, $existingCitations, $rawCitations]]
+     * @hook Citation::importCitations::after [[$publicationId, $existingCitations, $importedCitations]]
+     */
+    public function importCitations(int $publicationId, ?string $rawCitationList): void
+    {
+        $existingCitations = $this->getByPublicationId($publicationId);
+
+        Hook::call('Citation::importCitations::before', [$publicationId, $existingCitations, $rawCitationList]);
+
+        // Tokenize raw citations
+        $citationTokenizer = new CitationListTokenizerFilter();
+        $citationStrings = $rawCitationList ? $citationTokenizer->execute($rawCitationList) : [];
+
+        $existingRawCitations = collect($this->getByPublicationId($publicationId))
+            ->map(fn(Citation $citation) => $citation->getData('rawCitation'))
+            ->all();
+
+        if ($existingRawCitations !== $citationStrings) {
```

Comment:
Are the raw citations here the same? -- Because in $citation->setRawCitation we use cleanCitationString(). Is this fully the same what we get from the filter?